### PR TITLE
Swap NetworkPolicy storage to networking.k8s.io/v1

### LINKED ISSUE #50604: Switch storage version for networking.k8s.io/v1 and extensions/v1beta1's networking types to networking.k8s.io/v1
This includes updating `hack/test-update-storage-objects.sh` with a test to switch off `extensions/v1beta1` in order to verify that `networking.k8s.io/v1` keeps working, compare https://github.com/kubernetes/kubernetes/pull/50327#discussion_r132570895.

### PR DESCRIPTION
**Pull request by danwinship**
Finishes(?) the NetworkPolicy v1 migration.
Fixes #50604

The integration test passes. I copied the test-update-storage-objects.sh change from #50327 and have no idea if it's right.

/cc @sttts @caesarxuchao @thockin

**Release note**:
```release-note
```

### CODE REVIEW

**Comment by danwinship** [comment]
/test pull-kubernetes-e2e-kops-aws

everything else passed though I'm not sure what test, if any, runs test-update-storage-objects.sh...

File: cmd/kube-apiserver/app/server.go
Comment (thockin) [inline cmd/kube-apiserver/app/server.go]: Do we want a comment to remove this during v1.9 devel?

**Comment by thockin** [comment]
I have no context to review this, but I'm happy to approve when it is reviewed.  This missed the code freeze, so I assume it goes into 1.9, which means we can't remove the extensions path until 1.10?

File: cmd/kube-apiserver/app/server.go
Comment (sttts) [inline cmd/kube-apiserver/app/server.go]: I am nearly surprised that this change does not change the etcd prefix as well. But reading through https://github.com/kubernetes/kubernetes/blob/fc3b8bd58b0b20f549148fccbcbb0498f6f16f27/staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go#L336 this makes sense.

**Comment by sttts** [comment]
This obviously needs a release note.

I wonder how we deprecate the old extensions types. Do we have a good example to follow (e.g. extensions.HorizontalPodAutoscaler) ?

**Comment by sttts** [comment]
@liggitt @enj can you please take a look? Have we forgotten anything here?

File: hack/test-update-storage-objects.sh
Comment (enj) [inline hack/test-update-storage-objects.sh]: I am not familiar enough with this file to comment on if this makes sense.

File: cmd/kube-apiserver/app/server.go
Comment (enj) [reply cmd/kube-apiserver/app/server.go]: Along with the extensions type completely?

**Comment by enj** [comment]
> Finishes(?) the NetworkPolicy v1 migration

I assume you want to drop the extension type and all of its generated code at some point.

The changes to the etcd test look correct:

/approve

I will defer lgtm / final approval to someone who understands that BASH file.

File: cmd/kube-apiserver/app/server.go
Comment (liggitt) [reply cmd/kube-apiserver/app/server.go]: we shouldn't remove this until the extensions type is removed and migration is enforced. should networkpolicies be added to `update-storage-objects.sh`?

File: hack/test-update-storage-objects.sh
Comment (liggitt) [reply hack/test-update-storage-objects.sh]: I also have no idea about this file, though I did expect networkpolicies to be added to the list to migrate in update-storage-objects.sh

**Comment by sttts** [comment]
@wojtek-t you appear as author of many lines in hack/test-update-storage-objects.sh. Can you comment and potentially lgtm the changes for that file?

File: hack/test-update-storage-objects.sh
Comment (caesarxuchao) [reply hack/test-update-storage-objects.sh]: You also need to add an example networkingpolicy object at https://github.com/danwinship/kubernetes/blob/9468b49ad6e010aebaaeccaf412ba39f17726ff0/hack/test-update-storage-objects.sh#L112. Otherwise it's not tested.

Comment (wojtek-t) [reply hack/test-update-storage-objects.sh]: Yup - we shouldn't remove it and we should add an object of that type to actually test it.

Comment (danwinship) [reply hack/test-update-storage-objects.sh]: @wojtek-t we shouldn't remove *what*?

**Comment by danwinship** [comment]
> You also need to add an example networkingpolicy object at L112. Otherwise it's not tested.

OK, I've added an example object. How can I confirm that it's being tested (and passing)? Just running test-update-storage-objects.sh directly does not work. I eventually see:

    I0908 10:06:38.434883   28386 serve.go:85] Serving securely on 127.0.0.1:6443
    ...
    W0908 10:06:39.002823   28413 factory_object_mapping.go:418] Failed to download OpenAPI (Get https://127.0.0.1:8443/swagger-2.0.0.pb-v1: dial tcp 127.0.0.1:8443: getsockopt: connection refused), falling back to swagger

(note port mismatch)

**Comment by danwinship** [comment]
OK, pull-kubernetes-unit runs that test. And it fails:

    I0908 14:50:28.452] +++ [0908 14:50:28] Verifying networkpolicies/test/allow-from-red-to-blue has storage version v1beta1 in etcd
    ...
    W0908 14:50:28.583] !!! Error in ./hack/test-update-storage-objects.sh:151
    W0908 14:50:28.583]   Error in ./hack/test-update-storage-objects.sh:151. 'grep ${old_storage_version}' exited with status 0 1

File: hack/test-update-storage-objects.sh
Comment (caesarxuchao) [reply hack/test-update-storage-objects.sh]: @danwinship, could you confirm if this test catches the bug *after* you add a test type, but *without* fixing hack/upgrade-storage-objects.sh?

[edit] I see your latest comments confirmed the failure. That's good.

**Comment by danwinship** [comment]
@caesarxuchao the test currently appears to fail whether or not update-storage-objects.sh is updated. (Which is not surprising I guess since it's failing before update-storage-objects.sh gets run.)

**Comment by caesarxuchao** [comment]
@danwinship it looks like the test fails because the old objects is not in extensions/v1beta1. The test script [sets](https://github.com/danwinship/kubernetes/blob/6832457133580f11e760b437ea036fce69ed733a/hack/test-update-storage-objects.sh#L74) the storage version via the `--storage-version` flag. But according to the [code change](https://github.com/kubernetes/kubernetes/pull/51955/files#diff-36e1d7088209b1ab6ff878b18bb35210R569) you made, the storage version seems to only depend on the order of versions supplied to `AddCohabitatingResources`?

**Comment by thockin** [comment]
This didn't go in for 1.8, so we have to get it into 1.9 and wait for 1.10 to remove beta

**Comment by cmluciano** [comment]
@danwinship Can you rebase this?

**Comment by danwinship** [comment]
rebased. ftr I still have no clue how to fix the tests

**Comment by cmluciano** [comment]
I'm still trying this out locally as well. @caesarxuchao Any ideas on the failure report? When I try this on master it seems that the pre-existing tests fail silently, so I wonder what the expected results of this script are.

File: cmd/kube-apiserver/app/server.go
Comment (cmluciano) [reply cmd/kube-apiserver/app/server.go]: @danwinship extensions and network ordering should be swapped here to look as follows

`storageFactory.AddCohabitatingResources(extensions.Resource("networkpolicies"), networking.Resource("networkpolicies"))`

We should also add a comment as mentioned by @thockin . DEPRECATED in 1.9, removed in 1.10 or something similar.

File: hack/test-update-storage-objects.sh
Comment (cmluciano) [inline hack/test-update-storage-objects.sh]: The 3rd space here if for the namespace. Based on the test file, this should be set to default or the test file should add the test namespace.

**Review by cmluciano — DISMISSED** [review]
@danwinship 

These changes should fix the tests. Please add the deprecation note as well.

File: cmd/kube-apiserver/app/server.go
Comment (danwinship) [reply cmd/kube-apiserver/app/server.go]: `extensions`-first-then-`networking` is the way the code already is in 1.8. So this would make the PR a no-op (except for the changes to the tests).

Comment (liggitt) [reply cmd/kube-apiserver/app/server.go]: @cmluciano this PR is switching to write to etcd in networking.k8s.io version, which is what we want

File: cluster/update-storage-objects.sh
Comment (liggitt) [inline cluster/update-storage-objects.sh]: `networkpolicies.networking.k8s.io`

**Comment by liggitt** [comment]
unit test failure:
```
W1019 15:55:50.545] E1019 15:55:50.545005   17702 server.go:418] Failed to create clientset with KUBE_API_VERSIONS="v1,storage.k8s.io/v1beta1,extensions/v1beta1,networking.k8s.io/v1,storage.k8s.io/v1". KUBE_API_VERSIONS is only for testing. Things will break.
W1019 15:55:50.546] E1019 15:55:50.545984   17702 server.go:546] Failed to setup bootstrap token authenticator because the loopback clientset was not setup properly.
W1019 15:55:50.550] E1019 15:55:50.550316   17702 apiserver.go:159] Failed to create clientset with KUBE_API_VERSIONS="v1,storage.k8s.io/v1beta1,extensions/v1beta1,networking.k8s.io/v1,storage.k8s.io/v1". KUBE_API_VERSIONS is only for testing. Things will break.
```

File: cmd/kube-apiserver/app/server.go
Comment (cmluciano) [reply cmd/kube-apiserver/app/server.go]: Ah, misread the comment

**Comment by thockin** [comment]
Still failing unit

/retest

**Comment by thockin** [comment]
/lgtm
/approve

If test needs changes, ping me back to re-lgtm

**Comment by danwinship** [comment]
/approve cancel
to get fejta-bot to stop

/hold
because I need to push some debugging code (unless someone can tell me how to set up an environment to run this test locally; AFAICT it only gets run from `hack/jenkins/test-dockerized.sh`, which isn't referenced from any other file in the repo)

**Comment by danwinship** [comment]
lol, integration test flake before test-update-storage-objects ran.
/retest

**Comment by danwinship** [comment]
OK, I think that `AddCohabitatingResources` effectively overrides `--storage-versions` in such a way as to break `test-update-storage-objects.sh`. It seems that if we `AddCohabitatingResources` with `extensions` first, then newly-created NetworkPolicy objects are always created as `extensions/v1beta1`, even if `--storage-versions` specifies `networking.k8s.io` first. And if we `AddCohabitatingResources` with `networking` first, then newly-created NetworkPolicy objects are always created as `networking.k8s.io/v1`, even if `--storage-versions` specifies `extensions` first.

(The one existing test in `test-update-storage-objects.sh` is for a resource type that just changed versions within the same API group, so there's no use of `AddCohabitatingResources` there, which is why the test works for them.)

Or maybe that just means that there's something else we need to be changing in the API server other than the `AddCohabitatingResources` call?

**Comment by cmluciano** [comment]
@danwinship This mirrors what I tried locally as well. I'm not sure how we can get this to work given the constraints of those flag overrides and I could not find the example of how other cohabitating resources fixed this.

**Comment by thockin** [comment]
Is this still on hold?

**Comment by danwinship** [comment]
Well, I don't know if it does what it's supposed to do. The current branch passes tests, but that's because I just removed the test-update-storage-objects changes. I guess the fact that etcd_storage_path_test passes is a good sign.

**Comment by liggitt** [comment]
this LGTM. the etcd_storage_path test demonstrates both groups encode to the same location in the expected version. I wouldn't block on the update objects shell test shortcomings

**Pull request merged**
